instrumentation/grpc/lib/opentelemetry/instrumentation/grpc/patches/core_call.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module GRPC
+      module Patches
+        # Patches to prepend to the ::GRPC::CoreCall, which is a C extension.
+        module CoreCall
+          # This is the heart of sending / receiving GRPC messages on the wire.
+          # Even though it's a C extension, we can still intercept it here to
+          # handle the log message requirements of the OpenTelemetry semantic conventions.
+          def run_batch(ops)
+            if ops.keys.include?(::GRPC::Core::CallOps::SEND_MESSAGE)
+              OpenTelemetry::Trace.current_span.add_event(
```

Comment:
I wish I had a better grasp of the execution model here. I keep thinking back to the way that the Rack instrumentation keeps track of its own Span instead of using the "current" span in case users add custom interceptors that generate their own spans and I want to make sure the gRPC span is the one that is augmented with the event.

Reply:
> I wish I had a better grasp of the execution model here.

Me too, honestly. 🤣 This one is rough, for sure.

> I keep thinking back to the way that the Rack instrumentation keeps track of its own Span instead of using the "current" span in case users add custom interceptors that generate their own spans and I want to make sure the gRPC span is the one that is augmented with the event.

I think this is an excellent concern. I'm going to try and refactor it along those lines. I _think_ the scenario you described could happen if someone wrote their own interceptor, and that would definitely be a really upsetting / embarrassing bug for us. Not the first one I'd have written, either. 🤦 

I think what really makes this hard is that the execution model of the official GRPC library is just all over the place - and the bit we're discussing here is actually a C extension ... and a lot of work happens in that C extension, which just makes it all harder.

That said, I think I see a way to get to the state we want. 👍